processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java

```diff
@@ -749,7 +750,16 @@ public ConstantMappingBuilder selectionParameters(SelectionParameters selectionP
         public PropertyMapping build() {
             // source
             String sourceErrorMessagePart = "constant '" + constantExpression + "'";
-            Type sourceType = ctx.getTypeFactory().getType( String.class );
+
+            Type sourceType;
+            if ( !isConstantDirectAssignable( targetType, constantExpression ) ) {
+                // convert to string
+                sourceType = ctx.getTypeFactory().getType( String.class );
```

Comment:
knowing what I know now.. If the target is a primitive or boxed type and we use the default conversion fall-back, (e.g. Integer.toValue) that we used to, you will get runtime exceptions.. But it could be that there's a user defined mapping method that takes this into consideration.. We simply don't know.

Reply:
I'll do another update and use patternmatchig for hex,oct,bin .. that's far more easy and we can change when going to java8 only.

Reply:
Won't this break backwards compatibility? It might be directly assignable. However, up to now we would have called a user method for converting `String` into `int`. However, now we would directly assign it.

Reply:
don't think it breaks anything.. Suppose you would have a hex/oct/bin pattern, and try to run it through the parse function it fails runtime. That's what I found out. So.. you first need to remove all `0b` `0x` `0b` prefixes, underscores and `L,l,F,f,D,d` suffixes before you can call parse

Reply:
in short: next to having more efficient code, we have more possibilities (e.g use of bit/octal/byte-mask-patterns).